Add SolarEdge light integration

amccarthy opened the pull request:
Limited to on/off control.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This change enhances the existing SolarEdge integration with support for controlling devices connected to the SolarEdge inverter via ZigBee. Only on/off light support is implemented at this stage as that is the only type of device I have. Other devices can potentially be controlled by the same mechanism.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

amccarthy committed:
Add SolarEdge light integration

Limited to on/off control.

homeassistant commented:
Hi @amccarthy, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#38688).

Once you do that we will be able to review and accept this pull request.

Thanks!

ctalkington commented on homeassistant/components/solaredge/config_flow.py:
```diff
@@ -66,16 +84,32 @@ async def async_step_user(self, user_input=None):
+                                CONF_ACCESS_TOKEN: token,
+                            },
+                        )
+                    else:
```
```suggestion
```

ctalkington commented on homeassistant/components/solaredge/config_flow.py:
```diff
@@ -66,16 +84,32 @@ async def async_step_user(self, user_input=None):
+                            },
+                        )
+                    else:
+                        return self.async_create_entry(
```
```suggestion
                    return self.async_create_entry(
```

ctalkington commented on homeassistant/components/solaredge/config_flow.py:
```diff
@@ -66,16 +84,32 @@ async def async_step_user(self, user_input=None):
+                        )
+                    else:
+                        return self.async_create_entry(
+                            title=name, data={CONF_SITE_ID: site, CONF_API_KEY: api}
```
```suggestion
                        title=name, data={CONF_SITE_ID: site, CONF_API_KEY: api}
```

ctalkington commented on homeassistant/components/solaredge/config_flow.py:
```diff
@@ -66,16 +84,32 @@ async def async_step_user(self, user_input=None):
+                    else:
+                        return self.async_create_entry(
+                            title=name, data={CONF_SITE_ID: site, CONF_API_KEY: api}
+                        )
```
```suggestion
                    )
```

ctalkington commented on homeassistant/components/solaredge/__init__.py:
```diff
@@ -2,7 +2,7 @@
+from homeassistant.const import CONF_ACCESS_TOKEN, CONF_API_KEY, CONF_NAME
 import homeassistant.helpers.config_validation as cv
 from homeassistant.helpers.typing import HomeAssistantType
 
```
```suggestion
PLATFORMS = ["sensor", "light"]
```

ctalkington commented on homeassistant/components/solaredge/__init__.py:
```diff
@@ -40,4 +41,7 @@ async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry):
     hass.async_create_task(
```
```suggestion
    for component in PLATFORMS:
        hass.async_create_task(
            hass.config_entries.async_forward_entry_setup(entry, component)
        )
```

ctalkington commented on homeassistant/components/solaredge/__init__.py:
```diff
@@ -40,4 +41,7 @@ async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry):
     hass.async_create_task(
         hass.config_entries.async_forward_entry_setup(entry, "sensor")
```
```suggestion
```

ctalkington commented on homeassistant/components/solaredge/__init__.py:
```diff
@@ -40,4 +41,7 @@ async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry):
     hass.async_create_task(
         hass.config_entries.async_forward_entry_setup(entry, "sensor")
     )
```
```suggestion
```

ctalkington commented on homeassistant/components/solaredge/__init__.py:
```diff
@@ -40,4 +41,7 @@ async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry):
     hass.async_create_task(
         hass.config_entries.async_forward_entry_setup(entry, "sensor")
     )
+    hass.async_create_task(
```
```suggestion

```

ctalkington commented on homeassistant/components/solaredge/__init__.py:
```diff
@@ -40,4 +41,7 @@ async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry):
     )
+    hass.async_create_task(
+        hass.config_entries.async_forward_entry_setup(entry, "light")
+    )
```
```suggestion
```

ctalkington commented on homeassistant/components/solaredge/__init__.py:
```diff
@@ -40,4 +41,7 @@ async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry):
         hass.config_entries.async_forward_entry_setup(entry, "sensor")
     )
+    hass.async_create_task(
+        hass.config_entries.async_forward_entry_setup(entry, "light")
```
```suggestion
```

ctalkington commented on homeassistant/components/solaredge/__init__.py:
```diff
@@ -40,4 +41,7 @@ async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry):
+    hass.async_create_task(
+        hass.config_entries.async_forward_entry_setup(entry, "light")
+    )
     return True
```
```suggestion
    return True

async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    unload_ok = all(
        await asyncio.gather(
            *[
                hass.config_entries.async_forward_entry_unload(entry, component)
                for component in PLATFORMS
            ]
        )
    )

    return unload_ok
```

amccarthy committed:
Clean up SolarEdge light itegration add tests

amccarthy commented:
Pull request is still valid, waiting for review.

amccarthy committed:
Add device_info property to SolarEdge light

Cereal2nd commented:
I think this should be in th switch domain, and not in the light domain.

With this api you can control more then lights

frenck commented:
Sorry 🙏 , this one has been lingering around for a long time. Will be checking it quickly for some quick spots to get things rolling again...

amccarthy committed:
SolarEdge integration HA authentication

Authenticate with the HA service instead of using token extracted from
a browser session.

amccarthy commented on homeassistant/components/solaredge/config_flow.py:
```diff
@@ -66,7 +84,23 @@ async def async_step_user(self, user_input=None):
+                        )
                     return self.async_create_entry(
                         title=name, data={CONF_SITE_ID: site, CONF_API_KEY: api}
                     )
```
I can make this a single entry creation. I also am in the process of changing it to a username/password instead of extracting the cookie from a browser session.

frenck commented on homeassistant/components/solaredge/light.py:
```diff
@@ -0,0 +1,121 @@
+    await hass.async_add_executor_job(service.update)
+
+
+class SolarEdgeLightService:
```
This PR got merged today: #45734

It adds the DataUpdateCoordinator to part of this integration. Worth a look 😉

amccarthy replied:
I will look into using that class. The tutorial I was following to create the integration didn't mention it.

amccarthy commented on homeassistant/components/solaredge/__init__.py:
```diff
@@ -1,20 +1,25 @@
                 vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
                 vol.Required(CONF_API_KEY): cv.string,
                 vol.Required(CONF_SITE_ID): cv.string,
+                vol.Optional(CONF_ACCESS_TOKEN): cv.string,
```
Does that mean I can leave out the new configuration parameters from this section and then remove the existing ones in a separate PR?

frenck replied:
Yes

bdraco commented on homeassistant/components/solaredge/__init__.py:
```diff
@@ -1,20 +1,26 @@
                 vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
                 vol.Required(CONF_API_KEY): cv.string,
                 vol.Required(CONF_SITE_ID): cv.string,
+                vol.Optional(CONF_USERNAME): cv.string,
+                vol.Optional(CONF_PASSWORD): cv.string,
             }
```
Why do we need the username and password if we already have an api key?

amccarthy replied:
It is a different service that is not accessible with the api key.

Cereal2nd commented:
Why do we use the light domain and not the switch domain?

We can control a lot more then lights with this API.
https://www.solaredge.com/products/device-control#/

bdraco commented:
> Why do we use the light domain and not the switch domain?
> 
> We can control a lot more then lights with this API.
> [solaredge.com/products/device-control#](https://www.solaredge.com/products/device-control#/)

This seems like a completely different integration/service that uses an undocumented api and is only tangentially related to the solar monitoring api https://www.solaredge.com/sites/default/files/se_monitoring_api.pdf 

As the credentials are different, I think it make sense to create a new integration and not add it to the existing.  If solaredge suddenly blocks this api, it is going to leave users with half working config entries.

amccarthy committed:
Change SolarEdge light to use DataUpdateCoordinator

amccarthy committed:
Remove unnecessary config code
